<h3><b>Introduction to Data Analytics - Exercise set 2 - pandas-module</b></h3>

Put all your exercises (Jupyter Notebooks or Python-files) in your course Git-project.
Use either code comments or Jupyter Notebook markdown (text) to document which exercise you are doing and what a certain code section does! 
You can return all of these exercises in a single Jupyter Notebook, if you wish.

<b>The datasets for these exercises have been collected from kaggle.com<br />
(a service providing different datasets for practice)</b>

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_1.png" />

<h4><b>1. import pandas and read the csv-file found in Moodle (loans.csv). Use Python coding with pandas to answer the questions.</b></h4>


In [2]:
# you can import numpy and pandas here
import numpy as np
import pandas as pd

# and also load the needed csv-file
pd.options.display.float_format = '{:.2f}'.format
df = pd.read_csv('loans.csv')

C:\Users\marc\AppData\Local\Temp\ipykernel_19220\1191053091.py:3: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


<li>Remove the Customer ID –column from data</li>
<li>Print the head of the data</li>

In [26]:
# remove the column and print out the head of the DataFrame
df = df.drop('Customer ID', axis = 1)
df.head()

Loan ID Loan Status  Current Loan Amount  \
0  14dd8831-6af5-400b-83ec-68e61888a048  Fully Paid            445412.00   
1  4771cc26-131a-45db-b5aa-537ea4ba5342  Fully Paid            262328.00   
2  4eed4e6a-aa2f-4c91-8651-ce984ee8fb26  Fully Paid          99999999.00   
3  77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a  Fully Paid            347666.00   
4  d4062e70-befa-4995-8643-a0de73938182  Fully Paid            176220.00   

         Term  Credit Score  Annual Income Years in current job  \
0  Short Term        709.00     1167493.00              8 years   
1  Short Term           NaN            NaN            10+ years   
2  Short Term        741.00     2231892.00              8 years   
3   Long Term        721.00      806949.00              3 years   
4  Short Term           NaN            NaN              5 years   

  Home Ownership             Purpose  Monthly Debt  Years of Credit History  \
0  Home Mortgage   Home Improvements       5214.74                    17.20   
1  Home Mortgage  Debt Consolidation      33295.98                    21.10   
2       Own Home  Debt Consolidation      29200.53                    14.90   
3       Own Home  Debt Consolidation       8741.90                    12.00   
4           Rent  Debt Consolidation      20639.70                     6.10   

   Months since last delinquent  Number of Open Accounts  \
0                           NaN                     6.00   
1                          8.00                    35.00   
2                         29.00                    18.00   
3                           NaN                     9.00   
4                           NaN                    15.00   

   Number of Credit Problems  Current Credit Balance  Maximum Open Credit  \
0                       1.00               228190.00            416746.00   
1                       0.00               229976.00            850784.00   
2                       1.00               297996.00            750090.00   
3                       0.00               256329.00            386958.00   
4                       0.00               253460.00            427174.00   

   Bankruptcies  Tax Liens  
0          1.00       0.00  
1          0.00       0.00  
2          0.00       0.00  
3          0.00       0.00  
4          0.00       0.00

<li>Remove rows from the data that have <b style="color: red;">a too large of a loan</b> (Current Loan Amount should be less than 99999999)</li>

In [27]:
# remove loans that are too large
df = df[df['Current Loan Amount']<99999999]
df

Loan ID  Loan Status  Current Loan Amount  \
0      14dd8831-6af5-400b-83ec-68e61888a048   Fully Paid            445412.00   
1      4771cc26-131a-45db-b5aa-537ea4ba5342   Fully Paid            262328.00   
3      77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a   Fully Paid            347666.00   
4      d4062e70-befa-4995-8643-a0de73938182   Fully Paid            176220.00   
5      89d8cb0c-e5c2-4f54-b056-48a645c543dd  Charged Off            206602.00   
...                                     ...          ...                  ...   
99993  79b81158-5d55-4766-8ad6-ebcd683f7d59   Fully Paid             44484.00   
99994  8506a4e9-af7d-47d2-a1bf-7ea2c41858f0   Fully Paid            210584.00   
99995  3f94c18c-ba8f-45d0-8610-88a684a410a9   Fully Paid            147070.00   
99997  e1cb4050-eff5-4bdb-a1b0-aabd3f7eaac7   Fully Paid            103136.00   
99998  81ab928b-d1a5-4523-9a3c-271ebb01b4fb   Fully Paid            530332.00   

             Term  Credit Score  Annual Income Years in current job  \
0      Short Term        709.00     1167493.00              8 years   
1      Short Term           NaN            NaN            10+ years   
3       Long Term        721.00      806949.00              3 years   
4      Short Term           NaN            NaN              5 years   
5      Short Term       7290.00      896857.00            10+ years   
...           ...           ...            ...                  ...   
99993  Short Term        717.00     1152426.00            10+ years   
99994  Short Term        719.00      783389.00               1 year   
99995  Short Term        725.00      475437.00              7 years   
99997  Short Term        742.00     1150545.00              6 years   
99998  Short Term        746.00     1717524.00              9 years   

      Home Ownership             Purpose  Monthly Debt  \
0      Home Mortgage   Home Improvements       5214.74   
1      Home Mortgage  Debt Consolidation      33295.98   
3           Own Home  Debt Consolidation       8741.90   
4               Rent  Debt Consolidation      20639.70   
5      Home Mortgage  Debt Consolidation      16367.74   
...              ...                 ...           ...   
99993  Home Mortgage      small_business       6280.64   
99994  Home Mortgage               Other       3727.61   
99995       Own Home               other       2202.86   
99997           Rent  Debt Consolidation       7315.57   
99998           Rent  Debt Consolidation       9890.07   

       Years of Credit History  Months since last delinquent  \
0                        17.20                           NaN   
1                        21.10                          8.00   
3                        12.00                           NaN   
4                         6.10                           NaN   
5                        17.30                           NaN   
...                        ...                           ...   
99993                    21.00                         12.00   
99994                    17.40                         18.00   
99995                    22.30                           NaN   
99997                    18.80                         18.00   
99998                    15.00                           NaN   

       Number of Open Accounts  Number of Credit Problems  \
0                         6.00                       1.00   
1                        35.00                       0.00   
3                         9.00                       0.00   
4                        15.00                       0.00   
5                         6.00                       0.00   
...                        ...                        ...   
99993                     6.00                       0.00   
99994                     6.00                       0.00   
99995                     5.00                       0.00   
99997                    12.00                       1.00   
99998                     8.00                       0.00   

       Curr

<ul>
    <li>Remove rows that have the annual income as NaN (not a number)</li>
    <ul>
        <li><b>Extra task:</b> use imputation to use average income as the value instead of NaN</li>
    </ul>
</ul>

In [28]:
# remove rows if annual income is NaN
filtered = df[df['Annual Income'].notnull()]

#imputation
df['Annual Income'].fillna((df['Annual Income'].mean()), inplace=True)
df

C:\Users\marc\AppData\Local\Temp\ipykernel_15136\4084540400.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Annual Income'].fillna((df['Annual Income'].mean()), inplace=True)


Loan ID  Loan Status  Current Loan Amount  \
0      14dd8831-6af5-400b-83ec-68e61888a048   Fully Paid            445412.00   
1      4771cc26-131a-45db-b5aa-537ea4ba5342   Fully Paid            262328.00   
3      77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a   Fully Paid            347666.00   
4      d4062e70-befa-4995-8643-a0de73938182   Fully Paid            176220.00   
5      89d8cb0c-e5c2-4f54-b056-48a645c543dd  Charged Off            206602.00   
...                                     ...          ...                  ...   
99993  79b81158-5d55-4766-8ad6-ebcd683f7d59   Fully Paid             44484.00   
99994  8506a4e9-af7d-47d2-a1bf-7ea2c41858f0   Fully Paid            210584.00   
99995  3f94c18c-ba8f-45d0-8610-88a684a410a9   Fully Paid            147070.00   
99997  e1cb4050-eff5-4bdb-a1b0-aabd3f7eaac7   Fully Paid            103136.00   
99998  81ab928b-d1a5-4523-9a3c-271ebb01b4fb   Fully Paid            530332.00   

             Term  Credit Score  Annual Income Years in current job  \
0      Short Term        709.00     1167493.00              8 years   
1      Short Term           NaN     1373245.98            10+ years   
3       Long Term        721.00      806949.00              3 years   
4      Short Term           NaN     1373245.98              5 years   
5      Short Term       7290.00      896857.00            10+ years   
...           ...           ...            ...                  ...   
99993  Short Term        717.00     1152426.00            10+ years   
99994  Short Term        719.00      783389.00               1 year   
99995  Short Term        725.00      475437.00              7 years   
99997  Short Term        742.00     1150545.00              6 years   
99998  Short Term        746.00     1717524.00              9 years   

      Home Ownership             Purpose  Monthly Debt  \
0      Home Mortgage   Home Improvements       5214.74   
1      Home Mortgage  Debt Consolidation      33295.98   
3           Own Home  Debt Consolidation       8741.90   
4               Rent  Debt Consolidation      20639.70   
5      Home Mortgage  Debt Consolidation      16367.74   
...              ...                 ...           ...   
99993  Home Mortgage      small_business       6280.64   
99994  Home Mortgage               Other       3727.61   
99995       Own Home               other       2202.86   
99997           Rent  Debt Consolidation       7315.57   
99998           Rent  Debt Consolidation       9890.07   

       Years of Credit History  Months since last delinquent  \
0                        17.20                           NaN   
1                        21.10                          8.00   
3                        12.00                           NaN   
4                         6.10                           NaN   
5                        17.30                           NaN   
...                        ...                           ...   
99993                    21.00                         12.00   
99994                    17.40                         18.00   
99995                    22.30                           NaN   
99997                    18.80                         18.00   
99998                    15.00                           NaN   

       Number of Open Accounts  Number of Credit Problems  \
0                         6.00                       1.00   
1                        35.00                       0.00   
3                         9.00                       0.00   
4                        15.00                       0.00   
5                         6.00                       0.00   
...                        ...                        ...   
99993                     6.00                       0.00   
99994                     6.00                       0.00   
99995                     5.00                       0.00   
99997                    12.00                       1.00   
99998                     8.00                       0.00   

       Curr

<li>Get the average Current Loan Amount</li>

In [29]:
# get the average current loan
average = df['Current Loan Amount'].mean()
print(f"Average current loan: {average}")

Average current loan: 312313.59788060916


<li>Get the highest and lowest Annual Income in the dataset</li>

In [30]:
# get highest and lowest annual income in dataset
highest = df['Annual Income'].max()
lowest = df['Annual Income'].min()
print(f"Highest annual income: {highest} ")
print(f"Lowest annual income: {lowest} ")

Highest annual income: 165557393.0 
Lowest annual income: 76627.0 


<li>Get the Home Ownership value of the <b>Loan ID = bbf87a87-22cd-4d10-bd9b-7a9cc1b6e59d</b></li>

In [31]:
# get the needed value with the Loan ID
single = df[df['Loan ID'] == 'bbf87a87-22cd-4d10-bd9b-7a9cc1b6e59d']
single.iloc[0]['Home Ownership']

'Own Home'

<li>Create a new field into your dataset called <b>Actual Annual Income</b>.</li>
<br><b>Note:</b> The Actual Annual Income follow this formula:<br>
<b style="color: green">Annual Income – 12 * Monthly Debt</b><br><br>


In [32]:
# create a new field for Actual Annual income (annual income - 12 * monthly debt)
df['Actual Annual Income'] = df['Annual Income'] - (12 * df['Monthly Debt'])
df

Loan ID  Loan Status  Current Loan Amount  \
0      14dd8831-6af5-400b-83ec-68e61888a048   Fully Paid            445412.00   
1      4771cc26-131a-45db-b5aa-537ea4ba5342   Fully Paid            262328.00   
3      77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a   Fully Paid            347666.00   
4      d4062e70-befa-4995-8643-a0de73938182   Fully Paid            176220.00   
5      89d8cb0c-e5c2-4f54-b056-48a645c543dd  Charged Off            206602.00   
...                                     ...          ...                  ...   
99993  79b81158-5d55-4766-8ad6-ebcd683f7d59   Fully Paid             44484.00   
99994  8506a4e9-af7d-47d2-a1bf-7ea2c41858f0   Fully Paid            210584.00   
99995  3f94c18c-ba8f-45d0-8610-88a684a410a9   Fully Paid            147070.00   
99997  e1cb4050-eff5-4bdb-a1b0-aabd3f7eaac7   Fully Paid            103136.00   
99998  81ab928b-d1a5-4523-9a3c-271ebb01b4fb   Fully Paid            530332.00   

             Term  Credit Score  Annual Income Years in current job  \
0      Short Term        709.00     1167493.00              8 years   
1      Short Term           NaN     1373245.98            10+ years   
3       Long Term        721.00      806949.00              3 years   
4      Short Term           NaN     1373245.98              5 years   
5      Short Term       7290.00      896857.00            10+ years   
...           ...           ...            ...                  ...   
99993  Short Term        717.00     1152426.00            10+ years   
99994  Short Term        719.00      783389.00               1 year   
99995  Short Term        725.00      475437.00              7 years   
99997  Short Term        742.00     1150545.00              6 years   
99998  Short Term        746.00     1717524.00              9 years   

      Home Ownership             Purpose  Monthly Debt  \
0      Home Mortgage   Home Improvements       5214.74   
1      Home Mortgage  Debt Consolidation      33295.98   
3           Own Home  Debt Consolidation       8741.90   
4               Rent  Debt Consolidation      20639.70   
5      Home Mortgage  Debt Consolidation      16367.74   
...              ...                 ...           ...   
99993  Home Mortgage      small_business       6280.64   
99994  Home Mortgage               Other       3727.61   
99995       Own Home               other       2202.86   
99997           Rent  Debt Consolidation       7315.57   
99998           Rent  Debt Consolidation       9890.07   

       Years of Credit History  Months since last delinquent  \
0                        17.20                           NaN   
1                        21.10                          8.00   
3                        12.00                           NaN   
4                         6.10                           NaN   
5                        17.30                           NaN   
...                        ...                           ...   
99993                    21.00                         12.00   
99994                    17.40                         18.00   
99995                    22.30                           NaN   
99997                    18.80                         18.00   
99998                    15.00                           NaN   

       Number of Open Accounts  Number of Credit Problems  \
0                         6.00                       1.00   
1                        35.00                       0.00   
3                         9.00                       0.00   
4                        15.00                       0.00   
5                         6.00                       0.00   
...                        ...                        ...   
99993                     6.00                       0.00   
99994                     6.00                       0.00   
99995                     5.00                       0.00   
99997                    12.00                       1.00   
99998                     8.00                       0.00   

       Curr

<li>Get the Actual Annual Income of the loan with the <b>ID = 76fa89b9-e6a8-49af-afa1-8151315aba8e</b></li>

In [33]:
# get the Actual Annual income for the needed ID
single = df[df['Loan ID'] == '76fa89b9-e6a8-49af-afa1-8151315aba8e']
single.iloc[0]['Actual Annual Income']

674234.76

<li>Get the Loan ID of the loan with the smallest Actual Annual Income </li>

In [34]:
# get loan ID of lowest actual annual income
single = df[df['Actual Annual Income'] == df['Actual Annual Income'].min()]
single.iloc[0]['Loan ID']


'0a84b8ab-ed2f-474f-a3a9-8bf5bb90c31a'

<li>How many loans are "Long term"?</li>

In [35]:
# produce code that shows how many loans are long term
long_term = df[df['Term'] == 'Long Term']
print(f"{long_term['Term'].count()} loans are long term")                                


25475 loans are long term


<li>How many loaners have more than 1 bankruptcy?</li>

In [36]:
# produce code that shows how many loaners have more than 1 bankruptcy
multiple_bankruptcies = df[df['Bankruptcies'] > 1]
print(f"{multiple_bankruptcies['Bankruptcies'].count()} have more than 1 bankruptcy")    

476 have more than 1 bankruptcy


<li>How many Short Term loans are for Home Improvements?</li>

In [37]:
# produce code that shows how many of the Short Term loans are for Home Improvements
short_term_improvements = df[(df['Term'] == 'Short Term') & (df['Purpose'] == 'Home Improvements')]
print(f"{short_term_improvements['Purpose'].count()} short term loans are for Home Improvements") 

3490 short term loans are for Home Improvements


<li>How many unique loan purposes are there?</li>

In [38]:
# produce code that shows how many unique loan purposes there are
unique_purposes = df['Purpose'].unique()
print(f"there are {len(unique_purposes)} unique loan purposes")

there are 16 unique loan purposes


<li>What are the 3 most common loan purposes?</li>

In [39]:
# produce code that shows the 3 most common loan purposes
top_three_purposes = df['Purpose'].value_counts().head(3)
top_three_purposes

Purpose
Debt Consolidation    69506
other                  5409
Home Improvements      5124
Name: count, dtype: int64

<li>Is there a correlation between <b>Annual Income</b> and <b>Number of Open Accounts</b> or is there a correlation between <b>Number of Credit Problems</b> and <b>Bankruptcies</b>?</li>

In [41]:
# produce code that shows which correlation statement is true
# df['Annual Income'].corr(df['Number of Open Accounts'])
df['Number of Credit Problems'].corr(df['Bankruptcies'])

0.7507485199510444

<p>The correllation between <b>Number of Credit Problems</b> and <b>Bankruptcies</b> has a value of 0.75 so they have a close correllation.</p>

<h4><b>2. Download the purchases.csv from Moodle, and do the following observations:</b></h4>


<li>What was the total price sum of the Purchase Order Number 018H2015? (14 rows in total)</li>

In [121]:
# produce the code that gets the needed total price sum
df2 = pd.read_csv('purchases.csv')
order_df = df2[df2['Purchase Order Number'] == '018H2015']
order_df['Total Price'].sum()

1245.0400000000002

<li>What is the name and description of the purchased item with the Purchase Order Number 3176273?</li>

In [122]:
# produce the code that gets the needed name and description
order2_df = df2[df2['Purchase Order Number'] == '3176273']
name = order2_df.iloc[0]['Item Name']
description = order2_df.iloc[0]['Item Description']
print(f"NAME: {name}\nDESCRIPTION: {description}")

NAME: PC Desktop
DESCRIPTION: HP Prodesk 600


<li>How many occasions (rows) of purchase data happened during the year 2013?</li>

In [123]:
# produce the code that filters the needed data for year 2013
orders_2023 = df2[df2['Purchase Date'].str.contains("2013")]

# First element in shape is rows
rows = orders_2023.shape[0]
print(f"{rows} occasions of purchase data happened during the year 2013")

34024 occasions of purchase data happened during the year 2013


<ul><li>What are the 5 most common Departments in the data?</li>
<ul><li><b>Extra task: </b>What are 3 Departments using most money in the data?</li></ul>
</ul>

In [124]:
# produce the code that gtest the needed most common departments
departments = df2['Department Name']
counted_departments = departments.value_counts()
counted_departments.head(5)

Department Name
Corrections and Rehabilitation, Department of    10891
Correctional Health Care Services                 6593
Water Resources, Department of                    4582
Forestry and Fire Protection, Department of       4526
State Hospitals, Department of                    4281
Name: count, dtype: int64

In [125]:
# 3 Departments using most money in the data
department_money = df2.groupby('Department Name')['Total Price'].sum()
sorted_by_money = department_money.sort_values(ascending=False)
sorted_by_money.head(3)

Department Name
Health Care Services, Department of   21866849702.63
Public Health, Department of           2748240333.99
Social Services, Department of         1291192293.65
Name: Total Price, dtype: float64

<li>Sort the data by Department Name</li>

In [126]:
# sort the data by department name
df2.sort_values('Department Name')

Creation Date Purchase Date Fiscal Year LPA Number  \
6        11/30/2012    06/15/2012   2012-2013        NaN   
519      11/30/2012    06/15/2012   2012-2013        NaN   
20018    11/30/2012    06/15/2012   2012-2013        NaN   
31381    02/21/2013    03/20/2013   2012-2013        NaN   
27820    01/03/2013    02/14/2011   2012-2013        NaN   
...             ...           ...         ...        ...   
24826    12/17/2012    12/15/2012   2012-2013        NaN   
49068    06/13/2013    06/12/2013   2012-2013        NaN   
49069    03/26/2014    03/25/2014   2013-2014        NaN   
49057    06/13/2013    06/12/2013   2012-2013        NaN   
44469    05/07/2013    05/01/2013   2012-2013        NaN   

      Purchase Order Number  Requisition Number Acquisition Type  \
6                   11OAL07                 NaN     NON-IT Goods   
519                 11OAL09                 NaN     NON-IT Goods   
20018               11OAL10                 NaN     NON-IT Goods   
31381              11-12201                 NaN  NON-IT Services   
27820              11-12175                 NaN  NON-IT Services   
...                     ...                 ...              ...   
24826            4500192615                 NaN     NON-IT Goods   
49068            4500199013                 NaN         IT Goods   
49069            4500208992                 NaN     NON-IT Goods   
49057            4500199013                 NaN         IT Goods   
44469            4500197248                 NaN     NON-IT Goods   

         Acquisition Method                      Department Name  \
6                       NCB        Administrative Law, Office of   
519                     NCB        Administrative Law, Office of   
20018                   NCB        Administrative Law, Office of   
31381  Informal Competitive  African American Museum, California   
27820  Informal Competitive  African American Museum, California   
...                     ...                                  ...   
24826    Formal Competitive       Water Resources, Department of   
49068  Informal Competitive       Water Resources, Department of   
49069        State Programs       Water Resources, Department of   
49057  Informal Competitive       Water Resources, Department of   
44469        State Programs       Water Resources, Department of   

       Supplier Code  ... Classification Codes Normalized UNSPSC  \
6         1001493.00  ...             55101524       55101524.00   
519       1657720.00  ...             55101524       55101524.00   
20018     1657720.00  ...             55101524       55101524.00   
31381        7341.00  ...             82101801       82101801.00   
27820     1749414.00  ...             78141501       78141501.00   
...              ...  ...                  ...               ...   
24826       16299.00  ...   27131603\n40141605       27131603.00   
49068     1016373.00  ...             43232605       43232605.00   
49069     1743406.00  ...             15101506       15101506.00   
49057     1016373.00  ...             81112217       81112217.00   
44469     1743406.00  ...             78180301       78180301.00   

                              Commodity Title       Class  \
6                             Reference books 55101500.00   
519                           Reference books 55101500.00   
20018                         Reference books 55101500.00   
31381           Advertising campaign services 82101800.00   
27820             Freight forwarders services 78141500.00   
...                                       ...         ...   
24826                          Air regulators 27131600.00   
49068       Analytical or scientific software 43232600.00   
49069                      Gasoline or Petrol 15101500.00   
49057  Industry specific software maintenance 81112200.00   
44469                    Fuelling of vehicles 78180300.00   

                                             Class Title      Family  \
6                            

In [130]:
# Replace the $-sign of the Unit Price –column, and convert the value to a float
df2['Unit Price'] = df2['Unit Price'].str.replace('$', '').astype(float)
df2['Unit Price']

0           0.60
1       12587.05
2         199.00
3           4.95
4           1.00
          ...   
62652     100.50
62653     245.00
62654     245.00
62655   20475.00
62656    1181.44
Name: Unit Price, Length: 62657, dtype: float64

<ul>
<li><b>Small extra tasks for extra points</b></li>
<ul>
<li>How many purchases in the data were IT Goods and had the total price more than 50000 dollars?</li>
<li>How many of the purchases have anything to do with IT? (IT Goods, IT Services, IT Telecommunications)</li>

</ul>
</ul>

In [149]:
# produce the code for the extra tasks if you wish complete them
#  How many purchases in the data were IT Goods and had the total price more than 50000 dollars?
it_goods = df2[df2['Acquisition Type'] == 'IT Goods']
it_goods_over_5000 = it_goods[it_goods['Total Price'] > 5000].shape[0]
print(f'{it_goods_over_5000} of the purchases were IT Goods and had the total price more than 5000 dollars')

3141 of the purchases were IT Goods and had the total price more than 5000 dollars


In [150]:
# How many of the purchases have anything to do with IT? (IT Goods, IT Services, IT Telecommunications)
it_product_count = df2[df2['Acquisition Type'].str.contains('NON-IT') == False].shape[0]
print(f'{it_product_count} of the purchases had something to do with IT')

9558 of the purchases had something to do with IT


<ul>
<li><b>Other extra tasks for extra points</b></li>
<ul>
<li>Create a new DataFrame, where you have filtered out purchases that have a Total Price of 0 or less</li>
<li>For this DataFrame, use groupby() –function twice to group the purchases data by Acquisition Type, and then calculating the result first by sum() and then by mean()</li>
<li>Which two acquisition types have the largest sums and means after grouping the data?</li>
</ul>
</ul>

In [188]:
# produce the code for the extra tasks if you wish complete them
filtered_df2 = df2[df2['Total Price'] > 0]
sum_result = filtered_df2.groupby('Acquisition Type')['Total Price'].sum()
mean_result = filtered_df2.groupby('Acquisition Type')['Total Price'].mean()
print(f'SUM:\n{sum_result.to_string()}\n\nMEAN:\n{mean_result.to_string()}')

# create a dataframe from the sum_result and mean_result dataframes
result_df = pd.DataFrame({
    'Sum': sum_result,
    'Mean': mean_result
})

# ChatGPT:
# The 'Acquisition Type' is part of the result DataFrame (result_df) 
# because when you perform a groupby operation in Pandas, the grouping column(s) 
# become the index of the resulting DataFrame. 
print('\n\nResulting dataframe:')
result_df

SUM:
Acquisition Type
IT Goods                  213676581.04
IT Services              1156983637.17
IT Telecommunications        307903.39
NON-IT Goods             1034291622.78
NON-IT Services         30564202787.91

MEAN:
Acquisition Type
IT Goods                  28953.47
IT Services              682989.16
IT Telecommunications     51317.23
NON-IT Goods              26087.51
NON-IT Services         2481062.00


Resulting dataframe:


Sum       Mean
Acquisition Type                               
IT Goods                213676581.04   28953.47
IT Services            1156983637.17  682989.16
IT Telecommunications      307903.39   51317.23
NON-IT Goods           1034291622.78   26087.51
NON-IT Services       30564202787.91 2481062.00

In [189]:
# Two largest sums
result_df['Sum'].sort_values(ascending=False).head(2)

Acquisition Type
NON-IT Services   30564202787.91
IT Services        1156983637.17
Name: Sum, dtype: float64

In [190]:
# Two largest means
result_df['Mean'].sort_values(ascending=False).head(2)

Acquisition Type
NON-IT Services   2481062.00
IT Services        682989.16
Name: Mean, dtype: float64

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_2.png" />

<h4><b>3. Download the data_salaries_india.csv from Moodle, and consider the following questions of the data. Use any means in pandas (or even NumPy) you wish to explain your answers.</b></h4>


<ul>
<li>Before we can do anything with the salaries, we have to convert them into something more usable</li>
<ul>
<li>Note: the salaries can be yearly, monthly or hourly salaries</li>
<ul>
<li>We don't also need the Indian rupee –sign (₹)</li>
</ul>
<li>You can use the template in Moodle to help you out with this (Salary filtering, pandas exercise 3)</li>
</ul>
</ul>

In [4]:
# convert the salaries as needed
import pandas as pd

def yearly_wage(row):
    # the last two characters determine if it's yearly, monthly, hourly
    period = row['Salary'][-2:]
    
    # remove all commas and combine all numbers
    number = int(''.join(filter(str.isdigit, row['Salary'])))
    
    # if it's hourly, the average work hours per year in India is
    # approximately 2117.01 (might change in future)
    if period == "hr":
        number = int(number * 2117.01)
    elif period == "mo":
        # months to year
        number = int(number * 12)
    
    # return the yearly salary in integer format
    return number

df3 = pd.read_csv('data_salaries_india.csv')
df3['Salary'] = df3.apply(yearly_wage, axis=1)
#df3['Salary'] = df3['Salary'].str.replace('₹', '').str.replace(',', '').str.replace('/yr', '').astype('int')
df3

Company Name                  Job Title        Role  \
0                    Axis Bank             Data Scientist  Specialist   
1     Globex Digital Solutions  Machine Learning Engineer     Manager   
2           Tempworks Services             Data Scientist     Manager   
3                   BlueOptima  Machine Learning Engineer  Specialist   
4            Dell Technologies               Data Analyst  Specialist   
...                        ...                        ...         ...   
5122                   Rejolut  Machine Learning Engineer  Specialist   
5123     Bartleby Technologies  Machine Learning Engineer  Specialist   
5124                Clairvolex             Data Scientist     Manager   
5125       Piramal Enterprises               Data Analyst     Manager   
5126                  Unilever             Data Scientist  Specialist   

      Salaries Reported   Location   Salary  
0                     3  Bangalore  1171687  
1                     1  Hyderabad  3257305  
2                     1  Bangalore  5811205  
3                     2  Bangalore  1180965  
4                     5  New Delhi   602930  
...                 ...        ...      ...  
5122                  1     Mumbai   513690  
5123                  1  Bangalore  1309535  
5124                  1  Bangalore  4499222  
5125                  1     Mumbai  4612711  
5126                  2  Bangalore  1705561  

[5127 rows x 6 columns]

<ul>
<li>What are the most common values in different fields (Job Titles, Companies, Location)? <b>Based on the distribution, is the data balanced or not?</b></li>
<ul>
<li><b>Extra task:</b> there seem to be some Job Titles that are almost the same, like "Machine Learning Data Associate" and "Machine Learning Associate", combine these into something common</li>
</ul>
</ul>

In [31]:
# produce the code that shows the most common values in different fields
# and answer the question about the distribution
most_common_name = df3['Company Name'].value_counts().idxmax()
most_common_job = df3['Job Title'].value_counts().idxmax()
most_common_location = df3['Location'].value_counts().idxmax()

print(f"Most common company name: {most_common_name}")
print(f"Most common job title: {most_common_job}")
print(f"Most common location: {most_common_location}")

df3['Salary'].describe()


Most common company name: Tata Consultancy Services
Most common job title: Data Scientist
Most common location: Bangalore


count       5127.00
mean     1232570.45
std      1816491.84
min        10814.00
25%       462977.00
50%       823971.00
75%      1435941.00
max     74445327.00
Name: Salary, dtype: float64

<h4>Because Salary cannot be negative and the standard deviation is higher than the mean, the Salary distribution is skewed to the right</h4>

In [32]:
df3[df3['Job Title'] == 'Machine Learning Associate']

Company Name                   Job Title        Role  Salaries Reported  \
42       Amazon  Machine Learning Associate  Specialist                  6   

     Location  Salary  
42  Hyderabad  295114

In [43]:
# Combine "Machine Learning Data Associate" and "Machine Learning Associate" into one category
df3['Job Title'] = df3['Job Title'].replace('Machine Learning Data Associate', 'Machine Learning Associate')
df3['Job Title'] = df3['Job Title'].replace('Machine Learning Data Associate I', 'Machine Learning Associate')
df3['Job Title'] = df3['Job Title'].replace('Machine Learning Data Associate II', 'Machine Learning Associate')
df3[df3['Job Title'] == 'Machine Learning Associate']

Company Name                   Job Title        Role  Salaries Reported  \
42         Amazon  Machine Learning Associate  Specialist                  6   
236        Amazon  Machine Learning Associate  Specialist                  5   
689        Amazon  Machine Learning Associate  Specialist                 38   
1681       Amazon  Machine Learning Associate  Specialist                  1   
3543       Amazon  Machine Learning Associate  Specialist                  9   

       Location  Salary  
42    Hyderabad  295114  
236   Hyderabad  383213  
689   Hyderabad  275841  
1681  New Delhi  329439  
3543  Hyderabad  258596

<li>Are there any outliers in the data that might affect the averages negatively (certain salaries)? Manage the outliers as you best see fit (either remove them or leave them, based on your analysis) </li>

In [5]:
# produce the code that shows the potential outliers and take care of them as you see fit
import dtale
dtale.show(df3).open_browser()

<h4>From the D-tale heatmap, I can spot two outliers with Salary values of 74445327 ₹ and 35791950 ₹ on the high end. The closest salary to them is 19234356 ₹. On the low end, there is an outlier with a salary of 10814 ₹. The closest salary to it is 21628 ₹.</h4>

<ul>
<li>If we want to correlate upon categories (ordinal or binary), we need to use factorize(). <b>Factorize the Role-column, and add the new column to the DataFrame.</b></li>
<ul>
<li><b>Note:</b> using <b>factorize()</b> for nominal categories (Job Title, Location, Company Name) doesn't work well, because the numbers do not have any numeric magnitude. In other words, these categories don't measure anything, they just group data, so numerical comparison / correlation doesn't mean anything statistically.</li>
</ul>
</ul>

In [ ]:
# example:
# factorize Role level into numbers
# label1, unique1 = pd.factorize(df['Role'], sort=False)
# df['ManagerRole'] = label1

# produce the needed factorization for Role-column

<ul>
<li><b>Finally, check out the correlations. Does anything correlate with anything? Can we make any assumptions?</b></li>
<ul>
<li>Tip: When correlating against binary variables, sometimes the Spearman correlation might be more sensitive, in pandas:</li>
</ul>
</ul>

In [ ]:
# Remember: df.corr(method="spearman")

<b>Extra tasks for this dataset</b><br><br>

<ul>
<li>After all salaries have been converted to correct format by using the helper function (check Moodle), use quantiles and split the data into four different parts:</li>
<ul>
<li><span style="color: red;">Top 25%</span><br>=> quantile(0.75)</li>
<li><span style="color: red;">Above average, values between top 25-50%</span><br>
=> quantile(0.5)  - quantile(0.75)</li>
<li><span style="color: red;">Below average, values between top 50-75% </span><br>
=> quantile(0.25) – quantile(0.5)
</li>
<li><span style="color: red;">Bottom 25%</span><br>=> quantile(0.25)</li>
<br>
<li>What are the salary ranges for each quantile?</li><li>
See examples in pandas-materials on how to use quantiles
</li>
</ul>
</ul>
<hr>

<ul>
<li>Did you get any ideas how this data could be improved? Do we need some particular new data or some other operations on the data? Should we filter something out based on some other column?<br><br><b>Provide arguments for your answers in code comments.</b><br><br> </li>
<ul>
<li>Note: There are many good possible answers here!</li>
<li>Tip: How about replacing the "Salaries Reported" column with actual rows based on that number? Try doing this with the data!</li>
<li>Remember: This data only represents data engineering salaries based on selected Indian cities.
The world is a vast place :)
</li>
</ul>
</ul>

In [ ]:
# code for the extra tasks


<h2><b>Advanced extra tasks for extra points (varying challenges, some require Googling):</b></h2>

<ul>
<li><b>1. Data merge is a useful tool when you have multiple files of data that have the exact same structure.</b><br><br>Download the two csv-files from Moodle (videogames1.csv, videogames2.csv), and combine them into one Data Frame. Lastly, save the Data Frame into a new csv-file => combined.csv.</li>
</ul>

In [3]:
# code for the advanced extra task

<ul>
<li><b>2. Functions and lambdas allow us to extend the operations we wish to do to columns and rows in pandas. </b><br><br>For example, the built-in functions may not be enough in all cases. Use the data of exercise 1 (loans.csv), and create a new column called "Income Group" that holds a text value. <br><br>Create either a function or a lambda, that determines the True or False –value based on the Annual Income –column. Use the following table to create the values:</li>
</ul>

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_3.png"/>

<li>After creating the function/lambda, you can use it by using pandas' .apply() –function.<br><br></li>

<li>Finally, get the amount of rows grouped by each of the new Income Group –field values and print them out.</li> 


In [ ]:
# code for the advanced extra task

<ul>
<li><b>3.	Normalization allows us to convert the values of numeric columns to be between 0 and 1.</b> This is helpful when two different numbers seem to follow the same trend, but have completely different value ranges. For example, gold and silver prices tend to follow similar patterns, but their market worth is quite different. By using normalization, we can compare these trends more easily. <br>

Get historical prices of both gold and silver, and compare them without and with normalization. You can plot the prices by using df.plot() –function. Check the dataset in list in Moodle for some alternatives for gold and silver prices.
</li>
</ul>
<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_4.png" />

<ul>
<li><b>4.	Create an account to kaggle.com, and find any dataset that interests you. </b><br>

There's a list of possibly interesting datasets listed in Moodle as well.
<br><br>
<b>Try to find interesting features in data, in particular:</b>
</li>
<ul>
<li>Clean up data first (rows with too many NaN –values), values that are way too big or small, insignificant columns etc.)</li>
<li>You can create new columns as well if it seems suitable! (either by using functions or other means)</li>
<li>Interesting correlations (.corr() –function) and other interesting features in the data. Is something surprising in the data?</li>
<li><b>Note: </b>There are many ways on how to approach this exercise.</li>
</ul>
</ul>

<img src="http://srv.plab.fi/~tuomasv/data_analytics_2023_images/exercise_set_2/es2_5.png" />

In [ ]:
# do any Kaggle.com -related extra tasks in their own Jupyter notebooks for easier coding